### PR TITLE
Add BLS12-377 Precompiles

### PR DESCRIPTION
**Pull request by emaG3m**
https://eips.ethereum.org/EIPS/eip-2539

### CODE REVIEW

File: core/vm/contracts.go [hunk 1]
```diff
@@ -978,3 +986,333 @@ func (c *bls12381MapG2) Run(input []byte) ([]byte, error) {
...
+
+// bls12377G1Mul implements EIP-2537 G1Mul precompile.
+type bls12377G1Mul struct{}
+
+// RequiredGas returns the gas required to execute the pre-compiled contract.
+func (c *bls12377G1Mul) RequiredGas(input []byte) uint64 {
+	return params.Bls12377G1MulGas
+}
+
+func (c *bls12377G1Mul) Run(input []byte) ([]byte, error) {
```
Comment (emaG3m) [inline core/vm/contracts.go:1042]: Run interface is different in `go-ethereum`. 
`Run(input []byte, caller common.Address, evm *EVM, gas uint64) ([]byte, uint64, error)` ---> `Run(input []byte) ([]byte, error)`

**Comment by prestwich** [comment]
should link to 2539 above :)

File: core/vm/contracts.go (hunk 1 above)
Comment (prestwich) [reply core/vm/contracts.go:1042]: it should be fine. I think celo-blockchain somewhat polluted the interface to add some of our weirder precompiles 😬

File: core/vm/contracts.go [hunk 2]
```diff
@@ -978,3 +986,333 @@ func (c *bls12381MapG2) Run(input []byte) ([]byte, error) {
...
+	// Implements EIP-2537 G1Mul precompile.
```
Comment (prestwich) [inline core/vm/contracts.go:1043]: should be 2539.
2537 is the old precompile. Please review throughout :)

File: core/vm/contracts.go [hunk 3]
```diff
@@ -978,3 +986,333 @@ func (c *bls12381MapG2) Run(input []byte) ([]byte, error) {
...
+		discount = params.Bls12377MultiExpDiscountTable[k-1]
+	} else {
+		discount = params.Bls12377MultiExpDiscountTable[dLen-1]
+	}
+	// Calculate gas and return the result
+	return (uint64(k) * params.Bls12377G1MulGas * discount) / 1000
+}
+
+func (c *bls12377G1MultiExp) Run(input []byte) ([]byte, error) {
+	// Implements EIP-2537 G1MultiExp precompile.
```
Comment (prestwich) [inline core/vm/contracts.go:1093]: here too

File: params/protocol_params.go [hunk 4]
```diff
@@ -139,11 +139,21 @@ const (
 	Bls12381PairingPerPairGas uint64 = 23000  // Per-point pair gas price for BLS12-381 elliptic curve pairing check
 	Bls12381MapG1Gas          uint64 = 5500   // Gas price for BLS12-381 mapping field element to G1 operation
 	Bls12381MapG2Gas          uint64 = 110000 // Gas price for BLS12-381 mapping field element to G2 operation
+
```
Comment (prestwich) [inline params/protocol_params.go:142]: these can be populated from here:

https://github.com/ethereum/EIPs/pull/2999

File: params/protocol_params.go [hunk 5]
```diff
@@ -139,11 +139,21 @@ const (
...
+	Bls12377PairingBaseGas    uint64 = 0 // Base gas price for BLS12-377 elliptic curve pairing check
+	Bls12377PairingPerPairGas uint64 = 0 // Per-point pair gas price for BLS12-377 elliptic curve pairing check
 )
 
 // Gas discount table for BLS12-381 G1 and G2 multi exponentiation operations
 var Bls12381MultiExpDiscountTable = [128]uint64{1200, 888, 764, 641, 594, 547, 500, 453, 438, 423, 408, 394, 379, 364, 349, 334, 330, 326, 322, 318, 314, 310, 306, 302, 298, 294, 289, 285, 281, 277, 273, 269, 268, 266, 265, 263, 262, 260, 259, 257, 256, 254, 253, 251, 250, 248, 247, 245, 244, 242, 241, 239, 238, 236, 235, 233, 232, 231, 229, 228, 226, 225, 223, 222, 221, 220, 219, 219, 218, 217, 216, 216, 215, 214, 213, 213, 212, 211, 211, 210, 209, 208, 208, 207, 206, 205, 205, 204, 203, 202, 202, 201, 200, 199, 199, 198, 197, 196, 196, 195, 194, 193, 193, 192, 191, 191, 190, 189, 188, 188, 187, 186, 185, 185, 184, 183, 182, 182, 181, 180, 179, 179, 178, 177, 176, 176, 175, 174}
 
+// Gas discount table for BLS12-377 G1 and G2 multi exponentiation operations
+var Bls12377MultiExpDiscountTable = [128]uint64{}
```
Comment (prestwich) [inline params/protocol_params.go:155]: can be populated 

https://github.com/ethereum/EIPs/pull/2999

**Review by prestwich — CHANGES_REQUESTED** [review]

**Review by prestwich — APPROVED** [review]

**Pull request merged**